### ISSUE TITLE
Out-of-order-time-window feature and Thanos sidecar are incompatible (compaction issue)

### ISSUE THREAD

**Issue by ngc104** [prometheus/prometheus#13112]
## tl;dr

When Prometheus enables `out_of_order_time_window`, the Thanos sidecar does not send some blocs to its backend when some out-of-order metrics are received.

## How the problem happens

In a standard Prometheus + Thanos sidecar configuration, Prometheus will scrap metrics in an ordered way. It will generate blocs of 2h, not compacted (because I run with `--storage.tsdb.max-block-duration=2h --storage.tsdb.min-block-duration=2h`). Thanos sidecar will find blocs with `compaction.level=1` and upload them to it backend storage. Yeah, it works.

Now imagine that I get out-of-order errors. This mostly happens when remote-write is enabled. OK, Prometheus will drop out-of-order metrics, fill my log and... it still works. But I'm unhappy : some metrics are dropped.

Since 2.39, I can enable the new feature `out_of_order_time_window` : 
```
storage:
  tsdb:
    out_of_order_time_window: 600s
```

Nice : my metrics are no more dropped. But no blocs are uploaded to Thanos backend.
**Thanos sidecar is now broken !**

## Why does out_of_order_time_window break Thanos sidecar ?

Let's investigate...

- As described in [this Thanos doc](https://thanos.io/tip/components/compact.md/#warning-only-one-instance-of-compactor-may-run-against-a-single-stream-of-blocks-in-a-single-object-storage), Thanos sidecar will not upload compressed blocs.
- Before I enable `out_of_order_time_window` all my blocs had `compaction.level=1`
- I have compaction disabled with `--storage.tsdb.max-block-duration=2h --storage.tsdb.min-block-duration=2h`

But when I enable `out_of_order_time_window` I have some compacted blocs (`compaction.level=2`).

Thanos sidecar will not upload such blocs. This explains why it is broken.

## How can I have compacted blocs with disabled compaction ?

Let's read again how I disable compaction : 
```
When --storage.tsdb.min-block-duration and --storage.tsdb.max-block-duration have the same value, the compaction is disabled.
```

You can read this in many places.

Prometheus-Operator has a parameter `disableCompaction` with the [code below](https://github.com/prometheus-operator/prometheus-operator/blob/ba536405154d18f3a6f312818283d671182af6f3/pkg/prometheus/server/statefulset.go#L377C1-L384C3) :

```
	if disableCompaction {
		thanosBlockDuration := "2h"
		if thanos != nil {
			thanosBlockDuration = operator.StringValOrDefault(string(thanos.BlockDuration), thanosBlockDuration)
		}
		promArgs = append(promArgs, monitoringv1.Argument{Name: "storage.tsdb.max-block-duration", Value: thanosBlockDuration})
		promArgs = append(promArgs, monitoringv1.Argument{Name: "storage.tsdb.min-block-duration", Value: thanosBlockDuration})
	}
```

Now let's think again : why is compaction disabled when `--storage.tsdb.min-block-duration` and `--storage.tsdb.max-block-duration` have the same value ?

In fact, **compaction is never disabled**. Having the same value for both parameters will only make it useless to compact them. When the min bloc already has the same size as the max allowed, there will be no other bloc to compact into a bigger bloc. This explains why compaction seems to be disabled.

However, with `out_of_order_time_window`, you may have blocs with expected metrics and sometimes another bloc with same start and end date with out-of-order metrics. When Prometheus finds 2 blocs with the same start and end date, and *with compaction that cannot be disabled, it will compact them* !

This explains why I find some blocs with `compaction.level=2` even when I run with `--storage.tsdb.max-block-duration=2h --storage.tsdb.min-block-duration=2h`.

## How to fix

There are many ways to fix.

- don't send out of order metrics. Easy to say. Not so easy to do (and thanks again for the out-of-order-time-window feature)
- don't use Thanos sidecar. Prefer store with a remote-write backend (Thanos also have such a receiver)
- let's open an issue at Thanos' to enable concurrency write of compacted blocs (sidecar and compactor). Easy to open, harder to implement.
- here at Prometheus' : add a new option to explicitely disable compaction
- ...

I open this issue here for 2 reasons :
- because I wanted to document why some blocs are compacted even when compaction is (implicitely) disabled
- because I if I'm not wrong, I wanted to write that today, out-of-order-time-window feature and Thanos sidecar are incompatible
- and mainly because the fix **add a new option to explicitely disable compaction** is probably the easiest to implement.

### What did you see instead? Under which circumstances?

(mandatory field, but I already explained it above)

### System information

Kubernetes / Prometheus-Operator

### Prometheus version

```text
prometheus, version 2.42.0 (branch: HEAD, revision: 225c61122d88b01d1f0eaaee0e05b6f3e0567ac0)
  build user:       root@c67d48967507
  build date:       20230201-07:53:32
  go version:       go1.19.5
  platform:         linux/amd64

thanos, version 0.30.2 (branch: HEAD, revision: fe3f5d24192570038e9576307e1b31794920a1f3)
  build user:       root@6213df2115a5
  build date:       20230131-20:03:09
  go version:       go1.19.5
  platform:         linux/amd64
```

### Prometheus configuration file

```yaml
/bin/prometheus ...  --web.enable-remote-write-receiver --storage.tsdb.wal-compression --storage.tsdb.max-block-duration=2h --storage.tsdb.min-block-duration=2h

Configuration file :

...
storage:
  tsdb:
    out_of_order_time_window: 600s
```
```


### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

**Comment by ngc104** [comment]
May block #12631 ?

**Comment by yeya24** [comment]
There is already an issue on Thanos side https://github.com/thanos-io/thanos/issues/6462.

The workaround we can do at Thanos side, for example in sidecar is to enable upload compacted blocks. But you need to make sure there are no compacted blocks created due to reasons other than OOO.

**Comment by yeya24** [comment]
https://github.com/prometheus/prometheus/pull/13393
I think we should allow setting this configuration in Prometheus so that your Prometheus deployment doesn't compact OOO blocks with regular blocks and those blocks can be uploaded and handled by Compactor.

**Comment by yeya24** [comment]
Proposal: 
expose `EnableOverlappingCompaction` as a hidden flag in Prometheus and it defaults to true in Prometheus.
Thanos users can configure this hidden flag to false for this usecase.

**Comment by bboreham** [comment]
No objection to a hidden flag.

Why wouldn’t you make the flag “disable compaction” ?

**Comment by bboreham** [comment]
> Or you suggested for this flag to disable compaction entirely, not just disabling overlapping compaction?

That’s what I meant. It seems from the issue this is the requirement.

**Comment by yeya24** [comment]
I see. In Thanos sidecar mode, there is already a prerequisite to disable Prometheus block compaction by setting min and max block range to 2h. But this doesn't prevent vertical compaction of OOO blocks from happening so that's why the flag is for overlapping compaction.

For Thanos usecase, that's already good enough. Disabling compaction totally (except head compaction) could also work I think. But since it requires additional code changes I will skip it and just expose the flag.

**Comment by gillg** [comment]
Side question, do you see a way to fix the past on Thanos ?
I have a bucket full of data since 1 year, but not read by Thanos for a reason I don't understand.
All my blocks are marked :
```json
{
	"ulid": "01J45Q72ZR6W894T5MX2HW0PVW",
	"minTime": 1722470400000,
	"maxTime": 1722477600000,
	"stats": {
             ...
	},
	"compaction": {
		"level": 1,
		"sources": [
			"01J45Q72ZR6W894T5MX2HW0PVW"
		],
		"hints": [
			"from-out-of-order"
		]
	},
	"version": 1,
	"thanos": {
		"downsample": {
			"resolution": 0
		},
		"source": "sidecar",
               ...
	}
}
```

The hint "from-out-of-order" is coming from prometheus if I understand correctly and makes Thanos ignore these blocks. But can we just remove this hint and the index should contains the data or the think the data is "corrupted" from Thanos point of view ?

**Comment by yeya24** [comment]
Hey @gillg, sorry I missed your comment.

> The hint "from-out-of-order" is coming from prometheus if I understand correctly and makes Thanos ignore these blocks.

This is just a hint set by Prometheus. Thanos doesn't care what kind of hints you have so it should be able to read your blocks regardless the hint.

Can you create an issue at Thanos repository instead? Thanks
